src/proto/streams/counts.rs

```diff
@@ -87,19 +111,26 @@ impl Counts {
         F: FnOnce(&mut Self, &mut store::Ptr) -> U,
     {
         let is_counted = stream.is_counted();
+        let is_pending_reset = stream.is_pending_reset_expiration();
 
         // Run the action
         let ret = f(self, &mut stream);
 
-        self.transition_after(stream, is_counted);
+        self.transition_after(stream, is_counted, is_pending_reset);
 
         ret
     }
 
     // TODO: move this to macro?
-    pub fn transition_after(&mut self, mut stream: store::Ptr, is_counted: bool) {
+    pub fn transition_after(&mut self, mut stream: store::Ptr, is_counted: bool, is_reset_counted: bool) {
         if stream.is_closed() {
-            stream.unlink();
+            if !stream.is_pending_reset_expiration() {
```

Comment:
I **think** that I'm following this logic (and that it is correct), but it would be super helpful to add a comment describing how `transition_after` works now that it is getting pretty involved.